### DISCUSSION TITLE
Hard wiring on gamecube

### DISCUSSION THREAD

**Discussion by vaderag** [darthcloud/BlueRetro#52] [Q&A] [answered]
Thought occurred to me that soldering to the controller ports and running the wire inside the console you could fit an ESP32 in the modem gap underneath

Question tho, if you did this could you still use standard controllers in the port?
Thinking it through I can't see any reason why not (so long as you didn't connect another pad), but wondering if the 5v might be overloaded?

Anyone tried?

**Comment by darthcloud** [comment, marked as answer]
That wouldn't work because you would end-up with collision between BlueRetro and original controller on the data line.

Unless you power off BlueRetro or use a switch on the data line.

**Reply by vaderag** [reply]
Thanks. So would _some_ data be sent even if no controller was connected to Bluetooth?

If you were sticking a switch on, would it only be the data line that needed breaking?
(Or alternatively the power...)

Edit:
Guess you could use external USB for the blueretro and therefore switch that instead rather than soldering the 5v

**Reply by darthcloud** [reply]
Yes data is sent as soon console is power on. Bluetooth and wired are independent. This is required as not all games/software support controller hotplug. So blueretro simulated controller are present right away at boot.

Dont worry about power.

**Reply by vaderag** [reply]
If the blueretro didn't detect a connection on 1, would it automatically push the first connection to output 2?

Just thinking about the switching - basic would be all on all off (so for 4 player I'd need a 4PST/4PDT), but would be cool to be able to mix and match (i.e. if I had switch for each input could plausibly do 1P wired, 2P bluetooth but unless there is a soft pass the hard wiring would be super complex)

**Reply by darthcloud** [reply]
All outputs are independent so they are all active.

However if you use the universal firmware, the system detection only happen on port 1, so if port 1 is disconnect the other 3 will stay passive, so you could get away using only one switch.

GC specific firmware would require 4 switch.

**Reply by vaderag** [reply]
Hmm.. food for thought.
I think I'll just do all 4 for future proofing into a 4PDT

To clarify, if Controller 1 is **not** connected in GC Firmware, then does the first Bluetooth controller output on the P2 port, or would you need to connect two devices to control player 2?

**Reply by darthcloud** [reply]
The mapping between wireless controller and wired one is hardcoded. So you need to connect two controller for one to be map on port 2. But you can disconnect the first, and 2nd we keep being the 2nd.

**Reply by vaderag** [reply]
For anyone interested, decided to get a 4P4T switch which I'll drop in on the inputs. 
Then wire up so that
Position 1: 1,2,3,4 => BlueRetro
Position 2: 1 wired => 2,3,4 GC  BlueRetro 1,2,3
Position 3: 1,2 wired => 3,4 GC BlueRetro 1,2
Position 4: 1,2,3 wired => 4 GC BlueRetro 1

Will also drop in a toggle on the 5V to kill power entirely in the case of 4P wired (e.g. Donkey Konga!)

**Reply by vaderag** [reply]
@darthcloud just wanted to confirm some things as I'm about to get to soldering...

1. Do I need to connect up all the grounds from each of the cube ports or is it sufficient to have just one running to the ESP32?
2. Similarly, in a 4P build, is it necessary to wire up all the 5V to the ESP32 or is running a since 5V from the P1 port sufficient to provide the power?

Basically in a 4P setup wondering if it will work simply with 6 wires total? (4xdata for each port, 5v from P1 and GND from P1) ?

Thanks

**Reply by darthcloud** [reply]
Yes should be ok.

**Comment by vaderag** [comment]
Well, I did it! And it works!
Slight breakage on the 3D printed base since I have waaaay too much wire length which I'm trying to pack in there, so will reprint that and work out what to do about the wire, but it works!
Switch on the side kills the 5V allowing the front ports to be used :)

![image](https://user-images.githubusercontent.com/17175988/135439511-e7ce2b66-291d-41be-b113-39bc9bc19c3f.png)
![image](https://user-images.githubusercontent.com/17175988/135439610-938ad0e3-ae8e-4a1b-b2b6-0ca8a22c1d24.png)
![image](https://user-images.githubusercontent.com/17175988/135439537-c30b81bf-7223-4081-8567-a88b92453595.png)
![image](https://user-images.githubusercontent.com/17175988/135439662-2fe9f277-366a-4e5d-a0f7-d745ff84bcff.png)

**Reply by TheMikeBachmann** [reply]
I was just talking with a friend about doing something similar a few days ago. This looks great! Would you be willing to share the part number of the switch you used and the stl of the port cover you fit this into? Thanks either way for doing the legwork on this!

**Reply by vaderag** [reply]
Any switch will do. I had this one lying around in an old project. It simply kills the 5v
I actually ordered some more complex switches to allow some ports on and others off but ultimately decided I was over complicating things!

The STL I've put on Thingiverse already
https://www.thingiverse.com/thing:4978953

**Reply by TheMikeBachmann** [reply]
Ah ok that's what I was interested in was doing combinations of controllers. So you abandoned the 4 positions you listed above and just went with all or nothing? Thanks for the info!

**Reply by MarkAdamson** [reply]
Hi, I'm thinking of doing this, but I'm unclear on why the switch is necessary? I'm comparing your install to [this version](https://www.reddit.com/r/Gamecube/comments/pz65vl/clean_internal_blueretro_install/) on reddit.

That guy seems to say that there's no need to turn the esp32 on or off - if you have wired controllers plugged in, they will override any signals from blueretro - is this correct?

So if I were to install this internally, like the other guy - no switch, all hardwired - then decide I wanted to play WITHOUT a bluetooth controller - just a wired controller in port 1, vanilla-style - could I do that?

**Reply by darthcloud** [reply]
BlueRetro emulate all four controllers regardless of BT ctrl being connected or not.

If it work like that then its pure luck and not the result of an intended design.

The wired controller probably somehow drive the bus high very strongly and win most of the time against blueretro trying to pull bus low.

Good if it work well for him but at scale of many people to try this, results are likely to vary between gc version, wired controller used, esp32 board used, etc. But it could also turn out to be actually stable, who knows?

In any case, not the kind of setup i will promote, sound like trouble waiting to happen. Try at your own risk.

I do intend to better support internal install down the line, wired ctrl will be actively looked for by BlueRetro to avoid two device talking simultaneously and BT ctrl number will be assigned base on that.

**Reply by MarkAdamson** [reply]
Thanks for the response, that's good to know. I especially like the idea of Blueretro actively looking for wired controllers and assigning BT numbers based on it! Sounds tricky to implement though.

I'm going to try my luck with the internal mod and see how it goes - and not complain if it doesn't work, since it's against your advise. If it doesn't I can always back up and do the switched version, and if I fry my gamecube, it's my own dumb fault 😄

**Reply by darthcloud** [reply]
It's very unlikely to fry your GameCube, when I say risk its more like loosing in a Game due to BlueRetro glitching the wired ctrl comm randomly.

**Reply by vaderag** [reply]
Fwiw, I just tested and it does cause issues if you try and use a standard controller (wavebird in my case) without the toggle switch
Neither controller registers inputs (even if I don't connect a pad to Blueretro). The moment I disconnect the wavebird the Blueretro works. Same vice versa
That makes the toggle essential imho, and I think as @darthcloud says his experience is pure luck

**Reply by vaderag** [reply]
> Fwiw, I just tested and it does cause issues if you try and use a standard controller (wavebird in my case) without the toggle switch
> Neither controller registers inputs (even if I don't connect a pad to Blueretro). The moment I disconnect the wavebird the Blueretro works. Same vice versa
> That makes the toggle essential imho, and I think as @darthcloud says his experience is pure luck

Also just confirmed that it doesn't work with bongos either. Sends the player select all crazy
Turning off blue retro with the switch solves it, as does disconnecting the wired bongo, so definitely related to interference between the two. 
https://user-images.githubusercontent.com/17175988/137922793-60ba0dee-8e58-435a-9474-11a7b03df4b7.mp4

Recommended to add a switch :)

**Reply by MarkAdamson** [reply]
> It's very unlikely to fry your GameCube, when I say risk its more like loosing in a Game due to BlueRetro glitching the wired ctrl comm randomly.

No, I assumed not, just exaggerating for effect really.

@vaderag Damn, that's disappointing... I'm tempted to try anyway though. I've got the bits ready to go, and I'm a very impatient person 😜 I might luck out, and if it doesn't work, I'll be able to use BT controllers just until I can get the switch and 3d-printed cover bought.

**Reply by vaderag** [reply]
> > It's very unlikely to fry your GameCube, when I say risk its more like loosing in a Game due to BlueRetro glitching the wired ctrl comm randomly.
> 
> No, I assumed not, just exaggerating for effect really.
> 
> @vaderag Damn, that's disappointing... I'm tempted to try anyway though. I've got the bits ready to go, and I'm a very impatient person 😜 I might luck out, and if it doesn't work, I'll be able to use BT controllers just until I can get the switch and 3d-printed cover bought.

Sure, give it a go. 
I don't have any wired controllers to hand as loaned them to a mate, but will test myself with a standard controller when I get chance! 

Sounds like @darthcloud plans to better support internal installs in the future anyway so that will probably eliminate any issues if/when it comes - I wouldn't mind dropping the switch myself but given I use Donkey Konga a fair bit it's a no go for me at the moment! That said you could grab a switch for pennies somewhere close I'm sure!

**Reply by seewood** [reply]
ive made a little helper pcb for this project. it should prevent damage on the cube and make it a lot easier to connect player 1 - 4. if you want to use original controller on e.g. Player 2 or Player 4, just disconnect the jumper cable from the prebuild connector - no need of a switch or something. Just open the bottom 3d print and disconnect the not needed cable.

you can use such angled connectors: 
![image](https://user-images.githubusercontent.com/33197691/139218785-ec7137c4-d9e6-4fb7-a05e-0ba2fa1f7612.png)

but maybe we need to create a new 3d print to clip it over this helper pcb. (the "Gamecube_Serial_Port_ESP32_Cover.stl" file from the shared thingiverse link)

![BlueRetro_adapter_pcb](https://user-images.githubusercontent.com/33197691/139217914-74508270-8bd4-4d89-bd6f-037657905584.png)

**Reply by MarkAdamson** [reply]
Just thought I'd update this question with my own outcome. I did the unswitched, hardwired mod first of all. It worked well enough with wired controllers - when plugging a wired controller, it would override the signal from blueretro. However, it didn't work with the GCN/GBA link cable - the signals conflicted, making the GBA inoperable with the gamecube.

Added to that, I clearly did _something_ wrong in my initial mod, so that my port four stopped worked altogether - wired OR wireless. I don't know if I bridged a connection or what, I couldn't tell.

So in the end, I bought a new controller port board, a switch, and a modified, 3d printed serial port 1 cover, and did @vaderag's switched mod using my (slightly) improved soldering skills. It all works much better, and I'm very happy with it.
